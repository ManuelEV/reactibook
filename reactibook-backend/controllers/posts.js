
const { response } = require('express'); 

const Post = require('../models/Post');

const getPosts = async (req, res = response) => {

	const posts = await Post.find();

	res.json({
		ok: true,
		posts,
	});

}

const createPost = async (req, res = response) => {

	const post = new Post(req.body);

	console.log(post);

	try {
		const savedPost = await post.save();

		res.json({
			ok: true,
			post: savedPost,
		});

	} catch (error) {
		console.log(error);
		res.status(500).json({
			ok: false,
			msg: 'No se pudo crear el nuevo post',
		});
	}

}

const updatePost = async (req, res = response) => {

	const postId = req.params.id;

	try {
		const post = await Post.findById(postId);

		if (!post) {
			return res.status(400).json({
				ok: false,
				msg: 'No existe post con ese Id',
			});
		}

		const newPost = {
			...req.body
		}

		const updatedPost = await Post.findByIdAndUpdate(postId, newPost, {new: true});

		res.json({
			ok: true,
			post: updatedPost,
		});

	} catch (error) {
		console.log(error);
		res.status(500).json({
			ok: false,
			msg: 'No se pudo actualizar el post',
		});
	}

}

const deletePost = async (req, res = response) => {

	const postId = req.params.id;

	try {
		const post = await Post.findById(postId);

		if (!post) {
			return res.status(400).json({
				ok: false,
				msg: 'No existe post con ese Id',
			});
		}

		await Post.findByIdAndDelete( postId );

        res.json({ ok: true });

	} catch (error) {
		console.log(error);
		res.status(500).json({
			ok: false,
			msg: 'No se pudo eliminar el post',
		});
	}

}

module.exports = {
	getPosts,
	createPost,
	updatePost,
	deletePost,
}