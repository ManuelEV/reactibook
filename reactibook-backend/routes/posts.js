/*
	Posts routes
	/api/posts
*/

const { Router } = require('express');

const { check } = require('express-validator');

const { isDate } = require('../helpers/isDate');

const { validateFields } = require('../middlewares/validate-fields');

const { getPosts, createPost, updatePost, deletePost } = require('../controllers/posts');

const router = Router();



router.get('/', getPosts);



router.post('/',
 [
 	check('content', 'El título es obligatorio').not().isEmpty(),
   check('visibility', 'La visibilidad del post es obligatoria').not().isEmpty(),
 	check('date', 'Formato de fecha no válido').custom(isDate),
 	validateFields,
 ], 
 createPost);


router.put('/:id', updatePost);


router.delete('/:id', deletePost);


module.exports = router;