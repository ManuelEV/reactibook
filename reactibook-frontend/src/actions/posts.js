import Swal from 'sweetalert2';
import { fileUpload } from '../helpers/fileUpload';
import { loadPosts } from "../helpers/loadPosts";
import { postOptions, putOptions } from '../helpers/requestOptions';
import { types } from "../types/types";


export const startNewPost = (content, visibility) => {
    return async( dispatch, getState ) => {
        const { uid, name } = getState().auth;

        const { currentFileUrl } = getState().posts;

        const newPost = {
            author: {id: uid, name},
            content,
            visibility,
            date: new Date().getTime(),
            fileUrl: currentFileUrl,
        };
        
        const options = postOptions(newPost);

        const response = await fetch(`${process.env.REACT_APP_API_URL}/posts`, options);

        const body = await response.json();

        const newPostId = body.post.id;

        dispatch(addNewPost({id: newPostId, ...newPost}));

    }
}

export const addNewPost = (newPost) => ({
    type: types.postsAddNew,
    payload: {
        ...newPost
    }
})

export const startLoadingPosts = () => {
    return async (dispatch) => {
        const posts = await loadPosts();
        console.log('Loading posts... ', posts)
        dispatch(setPosts(posts));
    }
}

export const setPosts = (posts) => ({

    type: types.postsLoad,
    payload: posts,

})

export const startUpdatingPost = (id, data) => {
    return async(dispatch, getState) => {

        const options = putOptions(data);

        const response = await fetch(`${process.env.REACT_APP_API_URL}/posts/${id}`, options);

        const body = await response.json();

        console.log(body);

        const editedPost = body.post;

       dispatch(refreshPost(editedPost));
        Swal.fire('Saved', 'El post se guardó con éxito', 'success');
    }
}

export const refreshPost = (post) => ({
    type: types.postsUpdated,
    payload: post
});


export const updateCurrentFileUrl = (fileUrl) => ({
    type: types.postsCurrentFileUrl,
    payload: fileUrl
})


export const startUploading = (file) => {
    return async (dispatch, getState) => {

        Swal.fire({
            title: 'Uploading',
            text: 'Please wait',
            allowOutsideClick: false,
            showConfirmButton: false,
        });
        Swal.showLoading();

        const fileUrl = await fileUpload(file);

        dispatch(updateCurrentFileUrl(fileUrl));
        
        Swal.close();

        console.log(fileUrl);

    }
}




export const startDeleting = (id) => {
    return async (dispatch, getState) => {

        const options = {
            method: 'DELETE',
        };

        const response = await fetch(`${process.env.REACT_APP_API_URL}/posts/${id}`, options);

        const body = await response.json();

        console.log(body);

        dispatch(deletePost(id));

    }
}

export const deletePost = (id) => ({
    type: types.postsDelete,
    payload: id,
})


export const postLogout = () => ({
    type: types.postsLogoutCleaning,
})