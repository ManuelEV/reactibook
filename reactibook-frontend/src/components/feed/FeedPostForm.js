import React from 'react'

export const FeedPostForm = () => {
    return (
        <div className="w-full flex justify-center">
            <form className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
                rounded">
                <textarea 
                    className="w-full rounded flex items-start p-2 text-black resize-none"
                    type="textarea"
                    autoComplete="off"
                    placeholder="¿Qué está pasando?"
                    
                ></textarea>

                <div className="w-full flex row mt-2 mx-2">
                    <div className="w-1/3 flex items-start">
                        <input
                            className=""
                            type="file"
                            name="file"
                            accept="image/png, image/jpeg"
                        />
                    </div>
                    <div className="w-2/3 flex justify-end">
                        <select
                            className="mr-2 text-black"
                            name="postVisibility"
                        >
                            <option value="friends">Amigos</option>
                            <option value="public">Público</option>
                        </select>
                        <button
                            className="bg-primary-600 rounded px-2 text-white
                            hover:bg-primary-500"
                        >
                            publicar
                        </button>
                    </div>
                </div>
            </form>
        </div>
    )
}
