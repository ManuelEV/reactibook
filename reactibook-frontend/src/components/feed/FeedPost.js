import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import 'moment/locale/es';
import { startDeleting, startUpdatingPost } from '../../actions/posts';

export const FeedPost = ({id, content, date, fileUrl, author}) => {

    moment().locale('es');

    const postDate = moment(date);

    const { uid } = useSelector( state => state.auth );

    const dispatch = useDispatch();

    const [editionMode, setEditionMode] = useState(false);

    const [editedPostContent, setEditedPostContent] = useState(content);

    const handleInputChange = (e) => {
        setEditedPostContent(e.target.value);
    }

    const [formErrorMsg, setformErrorMsg] = useState('');

    const isFormValid = () => {

        if ( content === editedPostContent ) {
            setformErrorMsg('El contenido del post no debe ser el mismo que el anterior');
            return false;
        } else if ( editedPostContent.trim().length === 0 ) {
            setformErrorMsg('El contenido del post no debe estar vacío');
            return false;
        }          

        return true;
    }

    const handlePostUpdate = () => {
        if (isFormValid()) {
            dispatch(startUpdatingPost(id, {content: editedPostContent}));
            setEditionMode(false);
        }
    }

    const handlePostDelete = () => {
        dispatch(startDeleting(id));
    }

    return (
        <div className="w-full flex justify-center">
            <div className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
                rounded">
                <div 
                    className="flex row mb-2"
                >
                    <img 
                            className="h-14 w-14"
                            src="./assets/user.svg" alt="user_profile"
                    />
                    <div className="grid grid-cols-1 ml-3">
                        <div className="text-xl font-semibold">{author.name}</div>
                        <div className="text-sm">{postDate.calendar()}</div>
                    </div>
                </div>

                
                {
                    editionMode ? (
                        <div className="w-full">
                            <textarea 
                                className="w-full rounded flex items-start p-2 text-black resize-none"
                                type="textarea"
                                autoComplete="off"
                                placeholder="¿Qué está pasando?"
                                name="editedPostContent"
                                value={editedPostContent}
                                onChange={handleInputChange}
                            >
                            </textarea>
                            <span className="text-red-500 text-sm">
                                {formErrorMsg}
                            </span>
                            <div className="w-full flex justify-end px-3 text-sm underline">
                                <span 
                                    className="pr-2 cursor-pointer"
                                    onClick={() => setEditionMode(false)}
                                >
                                    Cancelar
                                </span>
                                <span
                                    className="cursor-pointer"
                                    onClick={handlePostUpdate}
                                >
                                    Guardar
                                </span>
                            </div>
                        </div>
                    )

                    : (
                        <div 
                            className="w-full rounded flex items-start p-2"
                        >
                            {content}
                        </div>
                    )
                }

                <div 
                    className="w-full flex justify-center"
                >
                    <img src={fileUrl} alt="" />
                </div>
                
                {
                    (uid === author.id) &&
                    (<div className="w-full flex items-start px-3 text-sm underline">
                        <span 
                            className="pr-2 cursor-pointer"
                            onClick={() => setEditionMode(true)}
                        >
                            Editar
                        </span>
                        <span
                            className="cursor-pointer"
                            onClick={handlePostDelete}
                        >
                            Eliminar
                        </span>
                    </div>)
                }
            </div>
        </div>
    )
}
