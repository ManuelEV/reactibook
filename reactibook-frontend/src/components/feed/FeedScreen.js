import React, { useState } from 'react'
import { useSelector } from 'react-redux';
import { NavBar } from '../navigation/NavBar'
import { FeedPost } from './FeedPost';
import { FeedPostForm } from './FeedPostForm'

export const FeedScreen = () => {

    const {posts} = useSelector( state => state.posts );

    const [filteredPosts, setPosts] = useState(posts);

    const handleFilterPosts = (filterCondition) => {
        setPosts( posts.filter(post => post.visibility === filterCondition) );
    }

    return (
        <div className="w-full h-full">

            <NavBar />
            
            <div className="pt-16">

                <FeedPostForm />

                <div className="w-full flex justify-center">
                    <div className="w-3/5 flex row">
                        <span 
                            className="px-2 underline cursor-pointer"
                            onClick={() => handleFilterPosts('public')}
                        >
                            PÃºblico
                        </span>
                        <span 
                            className="px-2 underline cursor-pointer"
                            onClick={() => handleFilterPosts('friends')}
                        >
                            Amigos
                        </span>
                        <span 
                            className="px-2 underline cursor-pointer"
                            onClick={() => setPosts(posts)}
                        >
                            Ver todo
                        </span>
                    </div>
                </div>

                {
                    filteredPosts.map(post => (
                        <FeedPost
                            key={post.id}
                            {...post}
                        />
                    ))
                }
            </div>

        </div>
    )
}
