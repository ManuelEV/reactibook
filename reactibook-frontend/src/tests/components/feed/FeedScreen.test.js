import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { FeedScreen } from '../../../components/feed/FeedScreen';
import { posts } from '../../data/posts';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {uid: 'id', name: 'test'},
    ui: {
        loading: false,
        msgError: null,
        loginErrors: {
            emailMsg: '',
            passwordMsg: '',
        },
    },
    posts: {
        posts: posts,
        currentFileUrl: ''
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount( 
    <Provider store={ store }>
        <FeedScreen /> 
    </Provider>
)

describe('Tests in <FeedScreen/>', () => {
    
    test('snapshot should match', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('the number of posts should match', () => {
        
        const postsCount = wrapper.find('FeedPost').length;

        expect(postsCount).toBe(posts.length);

    })
    

})

