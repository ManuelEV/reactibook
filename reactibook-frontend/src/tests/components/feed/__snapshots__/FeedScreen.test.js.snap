// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <FeedScreen/> snapshot should match 1`] = `
<div
  className="w-full h-full"
>
  <div
    className="flex flex-row p-4 bg-gray-700 w-full fixed
 bg-gradient-to-b from-primary-400 to-primary-600"
  >
    <div
      className="w-2/3 text-xl tracking-wide font-semibold"
    >
      <span
        className="cursor-pointer"
        onClick={[Function]}
      >
        REACTIBOOK
      </span>
    </div>
    <div
      className="w-1/3 flex justify-end"
    >
      <div
        className="bg-primary-600 px-2 py-1 mx-2"
      >
        test
      </div>
      <button
        className="bg-secondary-500 rounded font-bold px-2 py-1"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
  <div
    className="pt-16"
  >
    <div
      className="w-full flex justify-center"
    >
      <form
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
        onSubmit={[Function]}
      >
        <textarea
          autoComplete="off"
          className="w-full rounded flex items-start p-2 text-black resize-none"
          name="content"
          onChange={[Function]}
          placeholder="¿Qué está pasando?"
          type="textarea"
          value=""
        />
        <span
          className="text-sm text-red-500"
        />
        <div
          className="w-full flex row mt-2 mx-2"
        >
          <div
            className="w-1/3 flex items-start"
          >
            <input
              accept="image/png, image/jpeg"
              className=""
              id="fileSelector"
              name="file"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
            />
            <button
              className="bg-gray-200 text-black px-3 py-1 rounded"
              onClick={[Function]}
            >
              Subir imagen
            </button>
          </div>
          <div
            className="w-2/3 flex justify-end"
          >
            <select
              className="mr-2 text-black"
              name="visibility"
              onChange={[Function]}
              value="friends"
            >
              <option
                value="friends"
              >
                Amigos
              </option>
              <option
                value="public"
              >
                Público
              </option>
            </select>
            <button
              className="bg-primary-600 rounded px-2 text-white
 hover:bg-primary-500"
              type="submit"
            >
              publicar
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="w-3/5 flex row"
      >
        <span
          className="px-2 underline cursor-pointer"
          onClick={[Function]}
        >
          Público
        </span>
        <span
          className="px-2 underline cursor-pointer"
          onClick={[Function]}
        >
          Amigos
        </span>
        <span
          className="px-2 underline cursor-pointer"
          onClick={[Function]}
        >
          Ver todo
        </span>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Manuel Espinoza
            </div>
            <div
              className="text-sm"
            >
              ayer a las 15:58
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          bojack horseman
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src="https://res.cloudinary.com/manuelev/image/upload/v1623268724/qcwdk8we04sql2kdukza.jpg"
          />
        </div>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Manuel Espinoza
            </div>
            <div
              className="text-sm"
            >
              ayer a las 3:37
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          Probando post con imagen
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src="https://res.cloudinary.com/manuelev/image/upload/v1623224234/xnzken0blhp8thkgdgee.jpg"
          />
        </div>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Juanito
            </div>
            <div
              className="text-sm"
            >
              ayer a las 3:01
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          Mi primer post con imagen
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src="https://res.cloudinary.com/manuelev/image/upload/v1623222065/kjrn6fl0g7g9vf7o9j2y.png"
          />
        </div>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Juanito
            </div>
            <div
              className="text-sm"
            >
              ayer a la 1:57
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          Este post lo verán solo mis amigos
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src=""
          />
        </div>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Juanito
            </div>
            <div
              className="text-sm"
            >
              ayer a la 1:57
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          Soy nuevo en esta red! este es mi nuevo post
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src=""
          />
        </div>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Manuel Espinoza
            </div>
            <div
              className="text-sm"
            >
              el martes pasado a las 23:17
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          probando el segundo post
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src=""
          />
        </div>
      </div>
    </div>
    <div
      className="w-full flex justify-center"
    >
      <div
        className="bg-gray-600 w-3/5 h-full px-5 py-5 mt-4
 rounded"
      >
        <div
          className="flex row mb-2"
        >
          <img
            alt="user_profile"
            className="h-14 w-14"
            src="./assets/user.svg"
          />
          <div
            className="grid grid-cols-1 ml-3"
          >
            <div
              className="text-xl font-semibold"
            >
              Manuel Espinoza
            </div>
            <div
              className="text-sm"
            >
              el martes pasado a las 23:17
            </div>
          </div>
        </div>
        <div
          className="w-full rounded flex items-start p-2"
        >
          probando el primer post
        </div>
        <div
          className="w-full flex justify-center"
        >
          <img
            alt=""
            src=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
